### ISSUE TITLE
What software is been used by users 

### ISSUE THREAD

**Issue by Sabryr** [NAICNO/Jobanalyzer#20]
Parking another issue to comeback later

It will be useful to know which software the users are using. e.g. which modules they load.

**Comment by lars-t-hansen** [comment]
This will for sure require something new in sonar, or maybe it'll be a completely separate tool.  To the first order of approximation, everyone runs `python` and we want to know what they load.  We can in principle log the entire command line and not just the program name but:

- what they load may not be apparent on the command line
- the command line may contain some sensitive data, especially if we grab it before the program has a chance to censor it
- bjørn helge reports that there's a problem with monitoring software going into never-ending disk wait if it tries to read a program's full command line when the program is itself in disk wait (the system must be rebooted to get rid of the stuck process)

I think this use case will need tighter boundaries (are we interested only in software we manage or also in software the user herself manages?) and a very clear justification, as it will represent something very different from what we've been doing so far.

I suppose in principle we can enumerate all the modules in the system (hoping that nothing is privately installed...) and look at access times, but I'm not sure if that's what we want to see.

**Comment by lars-t-hansen** [comment]
Also see #92.  UiT will be working on this, they also have this use case.

**Comment by lars-t-hansen** [comment]
Also see #280, which has landed.  From this morning, this is a table ordered descending by the number of CPU seconds used by jobs on the ML nodes during the last 10 days:
```
$ sonalyze jobs -u- -f10d --fmt=awk,cmd,cputime/sec | \
    awk '{ time[$1] += $2 } END { for (i in time) print i, time[i] }' | \
    sort -k 2hr | \
    head -n 30
python3 35566737
python 33342215
linux_60.x86_64 18052134
python,python_<defunct> 14077692
scripts.train 8923206
MATLAB 2955183
Linus_FoodColle 1654242
glide_backend 1050801
perl,python3 848706
scripts.train,scripts.train_<defunct> 484848
gcs_server,python,ray::CPUActor,ray::IDLE,raylet 397302
wandb-service(2 390075
gdesmond 341034
jupyter-lab 293949
watch 287124
python,torchrun_<defunct>,wave 239700
python3,python3_<defunct> 210516
htop 120900
gdesmond,python3 99495
python3.9 97953
node 67131
code-1a5daa3a02,node 42363
glide_backend,perl 41319
gzip 25980
kited 25911
scripts.preproc 19371
julia 17274
jservergo 17145
tensorboard 17007
java 16323
```

**Comment by lars-t-hansen** [comment]
And Fox, for hack value (cutting off a little arbitrary), probably this represents about 7-8 days:
```
cp2k.popt 229796817
R 201276123
julia 173809173
R,rserver,rsession 26285640
phypsys 20292678
python,torchrun 19410624
l1002.exe,l1110.exe,l502.exe,l703.exe 12517698
python,wandb-service(2 8016354
pycompliantcomp,pycompliantcomp_<defunct>,python3,python3_<defunct> 7883790
python 5922531
git,git-gc-evoruns- 5556285
pycompliantcomp,python3,python3_<defunct> 4191294
python,python_<defunct>,wandb-service(2 3200796
python,python_<defunct>,slurm_script,wandb-service(2 3149688
l1002.exe,l1101.exe,l1110.exe,l502.exe,l703.exe 2709504
python3 1836654
l1002.exe,l1110.exe,l703.exe 1294398
python,python_<defunct> 885087
l401.exe,l502.exe,l701.exe,l703.exe 739914
phase_transitio 713220
l302.exe,l502.exe,l703.exe 701322
l301.exe,l502.exe,l703.exe 607620
l1002.exe,l703.exe 576795
python3,wandb-service(2 554004
TestBuild.x86_6,python3 417852
l1002.exe,l1110.exe,l401.exe,l502.exe,l703.exe 406320
l1002.exe,l1110.exe,l302.exe,l502.exe,l703.exe 321042
l1002.exe,l1110.exe,l302.exe,l601.exe,l703.exe 315486
l1002.exe,l1110.exe,l502.exe,l701.exe,l703.exe 307914
EvolvingModular,python3 305025
l1002.exe,l502.exe,l703.exe 284679
python3.10,torchrun_<defunct>,wave 269544
lmp 213288
l401.exe,l502.exe,l703.exe 187500
l1002.exe,l1110.exe,l303.exe,l502.exe,l703.exe 176757
l502.exe,l703.exe 165195
cc1plus,cicc,make,nvcc,ptxas 126663
jupyter-lab,python 119376
l502.exe,l702.exe,l703.exe 102729
l1002.exe,l401.exe,l703.exe 85050
pycompliantcomp,python3 52200
```

**Comment by lars-t-hansen** [comment]
> Just a note from the discussion with Stephan Oepen. "It would be useful to get the field of research recorded "

@Sabryr, I'm not quite sure what is meant by that?  The field of research as in how we and others determine the software that's run, or the field as in which software (and for what science) is run?

**Comment by Sabryr** [comment]
Some more to the wish list from LUMI meeting. 
- Modules are used less. Lot of user software and containers. It  would benefit if we can capture which libraries or packages the user process loading. Something like `lsof` level capture
- Is it possible to capture precession level of executed code. e.g. FP32 or FP64

**Comment by lars-t-hansen** [comment]
The number of FP instructions executed of various types is available through the intel performance counters (assuming Intel, of course - haven't checked AMD, probably is similar).  The data may be privileged.  The counters may have been disabled.  Obtaining the data may be costly.  Research is required.

Regarding modules, I think we'll want to look into what the module system offers in terms of logging.  ISTR there was something.

Regarding libraries, we can grub through /proc/pid/fd and look at open files (so, lsof), but this directory is privileged and so far we have been reluctant to give sonar root privileges.  Research is required.

Containers, see #300.

All of these require data beyond the data that are currently gathered.  As we've known for a while, sonar is but one example of a monitor that can feed useful data to jobanalyzer; these others are separate examples.  Given that we have a sensible component architecture now we can evolve the implementation to allow for multiple monitors.

**Comment by Sabryr** [comment]
It seems the UiO HPC-team is working on a module monitoring system. Better to get more details about there plans

**Comment by lars-t-hansen** [comment]
Yes, there have been some discussions, the latest in Åre.  A plan exists for how to get this information into Jobanalyzer.

**Comment by lars-t-hansen** [comment]
Specifically, there's a plan to hook into the slurm pre-run script to list the modules loaded at the time the job runs (probably as part of the sbatch script iirc?), along with other useful information at that point.  It simply becomes a part of the general slurm integration and data collection (#66).

**Comment by lars-t-hansen** [comment]
Axel is working on this, I've been chatting with him about it.  No conclusions yet.

**Comment by Sabryr** [comment]
I will have meetings with sonar team next week. May be I can chat with Axel as well. We need some idea about timeline of deliverables dependent on external contributors.
